jobsD.go

```diff
@@ -0,0 +1,519 @@
+			"JobRunRetriesOnErrorLimit": retryJobRun.RetriesOnErrorLimit,
+		}).Info("job run error retry has been queued")
+
+		j.Instance.JobRunsRequeuedError++
+		j.addJobRun(*retryJobRun)
+		return
+	}
+
+	rescheduleJobRun, rescheduleErr := jobRun.reschedule(j.db, j.Instance.ID, j.schedules)
+	if rescheduleErr != nil {
+		log.WithError(rescheduleErr).Error("could not reschedule job")
+		return
+	} else if rescheduleJobRun != nil {
+		j.Instance.JobRunsRescheduled++
+		j.addJobRun(*rescheduleJobRun)
+		return
+	}
+
+	if err := jobRun.close(j.db, j.Instance.ID); err != nil {
+		log.WithError(err).Error("could not close job run")
+	}
+}
+
+func (j *JobsD) jobResurrector() {
+	for {
+		select {
+		case <-j.producerCtx.Done():
+			j.producerCancelWait.Done()
+			return
+		case <-time.After(j.Instance.JobRetryCheck):
+			break
+		}
+		j.log.Trace("finding jobs to resurrect - started")
+
+		jobRuns := []JobRun{}
+		j.db.Where(
+			"run_started_by IS NOT NULL AND closed_by IS NULL AND retry_timeout_at <= ?",
+			time.Now(),
+		).Limit(j.Instance.JobPollLimit).Find(&jobRuns)
+
+		if len(jobRuns) > 0 {
+			j.log.WithField("count", len(jobRuns)).Debug("job runs for resurrection found")
+		}
+
+		for _, jobRun := range jobRuns {
+			log := jobRun.logger(j.log)
+
+			retryJobRun, retryErr := jobRun.retryOnTimeout(j.db, j.Instance.ID)
+			if retryErr != nil {
+				log.WithError(retryErr).Error("could not create job resurrection timeout retry")
+				continue
+			} else if retryJobRun != nil {
+				log.WithFields(logrus.Fields{
+					"JobRun.RetriesOnTimeoutCount": retryJobRun.RetriesOnTimeoutCount,
+					"JobRun.RetriesOnTimeoutLimit": retryJobRun.RetriesOnTimeoutLimit,
+				}).Info("job run timeout retry has been queued")
+
+				j.Instance.JobRunsRequeuedTimeout++
+				j.addJobRun(*retryJobRun)
+				continue
+			}
+
+			j.jobFinish(jobRun)
+		}
+
+		j.Instance.LastSeenAt = sql.NullTime{Valid: true, Time: time.Now()}
+		j.db.Save(j.Instance)
+
+		j.log.Trace("finding jobs to resurrect - completed")
+	}
+}
+
+func (j *JobsD) addJobRun(jobRun JobRun) {
+	j.runQ.Push(jobRun)
+
+	// Notify the delegator of the new item to run
+	// This needs to be async to prevent a job run from deadlocking trying to reschedule
+	go func() { j.runQNew <- struct{}{} }()
+}
+
+// Down shutsdown the JobsD service instance
+func (j *JobsD) Down() error {
+
+	j.log.Debug("shuting down the JobsD service - started")
+
+	j.producerCtxCancelFunc()
+	j.producerCancelWait.Wait()
+
+	j.workerCtxCancelFunc()
+	j.workertCxCancelWait.Wait()
+
+	close(j.runNow)
+	close(j.runQNew)
+	j.started = false
+
+	j.Instance.ShutdownAt = sql.NullTime{Valid: true, Time: time.Now()}
+	j.Instance.LastSeenAt = j.Instance.ShutdownAt
+	tx := j.db.Save(j.Instance)
+
+	j.log.Debug("shuting down the JobsD service - completed")
+
+	return tx.Error
+}
+
+// CreateRun .
+func (j *JobsD) CreateRun(job string, jobParams ...interface{}) *RunOnceCreator {
+	theUUID, err := uuid.NewUUID()
+	if err != nil {
+		panic(err)
```

Comment:
_panic_, really?

Reply:
Yes.. Its not obvious but the google UUID library only returns an error here if it cannot get the time. Since everything is very time based it would be wise to just die here.

I will add a comment as to why this is acceptable .